internal/receiver/smartagentreceiver/receiver.go

```diff
@@ -135,5 +149,62 @@ func (r *Receiver) createMonitor(monitorType string) (interface{}, error) {
 		output.AddExtraDimension(k, v)
 	}
 
-	return monitor, nil
+	// Configure SmartAgentConfigProvider to gather any global config overrides and
+	// set required envs.
+	configureEnvironmentOnce.Do(func() {
+		r.setUpSmartAgentConfigProvider(extensions)
+		setUpEnvironment()
+	})
+
+	// Note, that this receiver has to configure main collectd even for collectd/custom monitor,
+	// despite the fact that, that monitor stands up its own instance of collectd to prevent this
+	// panic "Main collectd instance should not be accessed before being configured".
+	if r.config.monitorConfig.MonitorConfigCore().IsCollectdBased() {
+		configureCollectdOnce.Do(func() {
+			r.logger.Info("Configuring collectd")
+			err = collectd.ConfigureMainCollectd(saConfigProvider.CollectdConfig())
+		})
+	}
+
+	return monitor, err
+}
+
+func (r *Receiver) setUpSmartAgentConfigProvider(extensions map[configmodels.Extension]component.ServiceExtension) {
+	// If smartagent extension is not configured, use the default config.
+	f := smartagentextension.NewFactory()
+	defaultCfg := f.CreateDefaultConfig().(smartagentextension.SmartAgentConfigProvider)
+	saConfigProvider = defaultCfg
+
+	// Do a lookup for any smartagent extensions to pick up common collectd options
+	// to be applied across instances of the receiver.
+	var foundAtLeastOne bool
+	var multipleSAExtensions bool
+	for c := range extensions {
+		if c.Type() != f.Type() {
+			continue
+		}
+
+		if foundAtLeastOne {
+			multipleSAExtensions = true
+			continue
+		}
+
+		var cfgProvider smartagentextension.SmartAgentConfigProvider
+		cfgProvider, foundAtLeastOne = c.(smartagentextension.SmartAgentConfigProvider)
+		if !foundAtLeastOne {
+			continue
+		}
+		saConfigProvider = cfgProvider
+		r.logger.Info("Smart Agent Config provider configured", zap.String("extension_name", c.Name()))
+	}
+
+	if multipleSAExtensions {
+		r.logger.Warn("multiple smartagent extensions found, using the first one encountered")
```

Comment:
We should probably keep track of and log the name of the extension we're using.

Reply:
There's an info log above in the for loop. Would that not sufficient?

Reply:
Right, I'm referring to the "using the first one encountered" statement, which is a bit of an implementation detail

Reply:
Updated.